src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -0,0 +1,599 @@
+        _api.invalidatedRows.y = _api.cellCount.y;
+    }
+    else
+    {
+        _api.invalidatedRows.x = 0;
+        _api.invalidatedRows.y = gsl::narrow_cast<u16>(clamp<int>(_api.invalidatedRows.y + delta, u16min, u16max));
+    }
+
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateAll() noexcept
+{
+    _api.invalidatedRows = invalidatedRowsAll;
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateCircling(_Out_ bool* const pForcePaint) noexcept
+{
+    RETURN_HR_IF_NULL(E_INVALIDARG, pForcePaint);
+    *pForcePaint = false;
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateTitle(const std::wstring_view proposedTitle) noexcept
+{
+    WI_SetFlag(_api.invalidations, ApiInvalidations::Title);
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateFont(const FontInfoDesired& fontInfoDesired, _Out_ FontInfo& fontInfo) noexcept
+{
+    return UpdateFont(fontInfoDesired, fontInfo, {}, {});
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateSoftFont(const gsl::span<const uint16_t> bitPattern, const SIZE cellSize, const size_t centeringHint) noexcept
+{
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateDpi(const int dpi) noexcept
+{
+    u16 newDPI;
+    RETURN_IF_FAILED(api_narrow(dpi, newDPI));
+
+    if (_api.dpi != newDPI)
+    {
+        _api.dpi = newDPI;
+        WI_SetFlag(_api.invalidations, ApiInvalidations::Font);
+    }
+
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateViewport(const SMALL_RECT srNewViewport) noexcept
+{
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::GetProposedFont(const FontInfoDesired& fontInfoDesired, _Out_ FontInfo& fontInfo, const int dpi) noexcept
+try
+{
+    // One day I'm going to implement GDI for AtlasEngine...
+    // Until then this code is work in progress.
+#if 0
+    wil::unique_hfont hfont;
+
+    // This block of code (for GDI fonts) is unfinished.
+    if (fontInfoDesired.IsDefaultRasterFont())
+    {
+        hfont.reset(static_cast<HFONT>(GetStockObject(OEM_FIXED_FONT)));
+        RETURN_HR_IF(E_FAIL, !hfont);
+    }
+    else if (requestedFaceName == DEFAULT_RASTER_FONT_FACENAME)
+    {
+        // GDI Windows Font Mapping reference:
+        // https://msdn.microsoft.com/en-us/library/ms969909.aspx
+
+        LOGFONTW lf;
+        lf.lfHeight = -MulDiv(requestedSize.Y, dpi, 72);
+        lf.lfWidth = 0;
+        lf.lfEscapement = 0;
+        lf.lfOrientation = 0;
+        lf.lfWeight = requestedWeight;
+        lf.lfItalic = FALSE;
+        lf.lfUnderline = FALSE;
+        lf.lfStrikeOut = FALSE;
+        lf.lfCharSet = OEM_CHARSET;
+        lf.lfOutPrecision = OUT_RASTER_PRECIS;
+        lf.lfClipPrecision = CLIP_DEFAULT_PRECIS;
+        lf.lfQuality = PROOF_QUALITY; // disables scaling for rasterized fonts
+        lf.lfPitchAndFamily = FIXED_PITCH | FF_MODERN;
+        // .size() only includes regular characters, but we also want to copy the trailing \0, so +1 it is.

```

Comment:
>so +1 it is. [](http://example.com/codeflow?start=95&length=12)

What +1?

Reply:
Well +1 length to copy, so that we not just include strlen() characters but also the terminating null byte.

Reply:
Yeah I just didn't see the arithmetic for +1 below here. It was just straight up values with no arithmetic.